node/rpc/src/chain_api.rs

```diff
@@ -50,6 +51,44 @@ where
     Ok(MessageJson(ret))
 }
 
+pub(crate) async fn chain_export<DB, B>(
+    data: Data<RPCState<DB, B>>,
+    Params(params): Params<ChainExportParams>,
+) -> Result<ChainExportResult, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    B: Beacon + Send + Sync + 'static,
+{
+    let (epoch, recent_roots, skip_old_msgs, out, TipsetKeysJson(tsk)) = params;
+
+    let chain_finality = data.state_manager.chain_config().policy.chain_finality;
+    let recent_roots = recent_roots.unwrap_or(chain_finality);
+    if recent_roots < chain_finality {
+        Err(&format!(
+            "\"recent-stateroots\" must be greater than {}",
+            chain_finality
+        ))?;
+    }
+
+    if recent_roots == 0 && skip_old_msgs {
+        Err("must pass recent-stateroots along with skip-old-messages")?;
+    }
+
+    let file = File::create(&out).await.map_err(JsonRpcError::from)?;
+    let writer = BufWriter::new(file.clone());
```

Comment:
why clone?

Reply:
I'm surprised clippy didn't bark at this. Fixed.